subprojects/docs/src/samples/userguide/files/fileCollections/kotlin/build.gradle.kts

```diff
@@ -0,0 +1,80 @@
+import java.nio.file.Paths
+
+// tag::simple-params[]
+val collection: FileCollection = layout.files(
+    "src/file1.txt",
+    File("src/file2.txt"),
+    listOf("src/file3.csv", "src/file4.csv"),
+    Paths.get("src", "file5.txt")
+)
+// end::simple-params[]
+
+file("src").mkdirs()
+file("src/dir1").mkdirs()
+file("src/file1.txt").mkdirs()
+file("src2").mkdirs()
+file("src2/dir1").mkdirs()
+file("src2/dir2").mkdirs()
+
+// tag::closure[]
+tasks.create("list") {
+    doLast {
+        var srcDir: File? = null
+
+        // Create a file collection using a closure
+        // TODO fix this, but how?
+        // once fixed, add expected-output-file: fileCollectionsWithClosure.out in fileCollectionsWithClosure.sample.conf
+        /*
+        val collection = layout.files({ srcDir.listFiles() })
+        srcDir = file("src")
+
+        println("Contents of ${srcDir.name}")
+        collection.map { relativePath(it) }.sorted().forEach { println(it) }
+
+        srcDir = file("src2")
+        println("Contents of ${srcDir.name}")
+        collection.map { relativePath(it) }.sorted().forEach { println(it) }
+        */
```

Comment:
Passing a Kotlin lambda is not supported, see https://github.com/gradle/kotlin-dsl/issues/1077

Passing a `Provider<T>` is supported for both Groovy and Kotlin DSLs:
```kotlin
val srcDir = file("src")

val collection = layout.files(provider {
    srcDir.listFiles()
})

println("Contents of ${srcDir.name}")
collection.map { relativePath(it) }.sorted().forEach { println(it) }

srcDir = file("src2")
println("Contents of ${srcDir.name}")
collection.map { relativePath(it) }.sorted().forEach { println(it) }
```

Let's use `provider {}` in Kotlin DSL samples onward until https://github.com/gradle/kotlin-dsl/issues/1077 gets fixed.

Reply:
I think the goal here is to show that the collection can be defined before the srcDir has an actual value, because it's evaluated lazily. And that each time we change the value of srcDir, evaluating it leads to a different result.

I changed the code to the following to make it compile and make it identical to the groovy code:

```
        var srcDir: File? = null

        val collection = layout.files(provider {
            srcDir?.listFiles()
        })

        srcDir = file("src")
        println("Contents of ${srcDir.name}")
        collection.map { relativePath(it) }.sorted().forEach { println(it) }

        srcDir = file("src2")
        println("Contents of ${srcDir.name}")
        collection.map { relativePath(it) }.sorted().forEach { println(it) }
```